Index: freelan.debian/src/configuration_helper.cpp
===================================================================
--- freelan.debian.orig/src/configuration_helper.cpp	2013-08-01 22:50:04.000000000 +0200
+++ freelan.debian/src/configuration_helper.cpp	2013-08-02 16:53:27.000000000 +0200
@@ -197,7 +197,6 @@
 
 void setup_configuration(fl::configuration& configuration, const boost::filesystem::path& root, const po::variables_map& vm)
 {
-	typedef boost::asio::ip::udp::resolver::query query;
 	typedef fl::security_configuration::cert_type cert_type;
 	typedef cryptoplus::pkey::pkey pkey;
 
